### PR TITLE
Average unpooling Tensor integration

### PR DESCRIPTION
**Pull request by Randl**
Also, test are fixed

### CODE REVIEW

File: test/test_average_unpooling_layer.h [hunk 1]
```diff
@@ -52,7 +52,9 @@ TEST(ave_unpool, forward) {
   l.bias_init(weight_init::constant(0.0));
   l.init_weight();
 
-  vec_t res = l.forward({{in}})[0][0];
+  std::vector<const tensor_t*> out;
+  l.forward({{in}}, out);
```
Comment (karandesai-96) [inline test/test_average_unpooling_layer.h]: `auto out = l.forward({{in}})` is now available, but you need to rebase.

File: test/test_average_unpooling_layer.h [hunk 2]
```diff
@@ -70,18 +72,20 @@ TEST(ave_unpool, forward_stride) {
...
     };
   // clang-format on
 
   l.weight_init(weight_init::constant(1.0));
   l.bias_init(weight_init::constant(0.0));
   l.init_weight();
 
-  vec_t res = l.forward({{in}})[0][0];
+  std::vector<const tensor_t*> out;
+  l.forward({{in}}, out);
```
Comment (karandesai-96) [inline test/test_average_unpooling_layer.h]: `auto out = l.forward({{in}})` here too.

File: tiny_dnn/layers/average_unpooling_layer.h [hunk 3]
```diff
@@ -30,27 +30,29 @@ inline void tiny_average_unpooling_kernel(
   std::vector<typename partial_connected_layer::wi_connections> &out2wi) {
   CNN_UNREFERENCED_PARAMETER(scale_factor);
   for_i(parallelize, in_data[0]->size(), [&](size_t sample) {
-    const vec_t &in = (*in_data[0])[sample];
-    const vec_t &W  = (*in_data[1])[0];
-    const vec_t &b  = (*in_data[2])[0];
-    vec_t &out      = (*out_data[0])[sample];
+    const Tensor<> in((*in_data[0])[sample]);
+    const Tensor<> W((*in_data[1])[0]);
+    const Tensor<> b((*in_data[2])[0]);
```
Comment (karandesai-96) [inline tiny_dnn/layers/average_unpooling_layer.h]: `average_unpooling_layer` does not have `has_bias` facility ?

File: tiny_dnn/layers/average_pooling_layer.h [hunk 4]
```diff
@@ -41,8 +41,9 @@ inline void tiny_average_pooling_kernel(
       float_t bias   = b.host_at(d);
       for (size_t i = 0; i < oarea; ++i, ++idx) {
         const auto &connections = out2wi[idx];
-        float_t value = 0.0;
-        for (auto connection : connections) value += in.host_at(connection.second);
+        float_t value           = 0.0;
+        for (auto connection : connections)
+          value += in.host_at(connection.second);
```
Comment (edgarriba) [inline tiny_dnn/layers/average_pooling_layer.h]: @Randl can we just have `in.at(...)` and let tensor decide whether use host or device ?

Comment (Randl) [inline tiny_dnn/layers/average_pooling_layer.h]: Good idea!

**Pull request merged**